### PR TITLE
Fix dynamic class assembly name

### PR DESCRIPTION
**Pull request by lzybkr**
Using the assembly name to hint at the source of the classes was
problematic in multiple ways.

This change stores the actual filename in an attribute on the assembly.

So for a given type, one can get the assembly this way:

[SomeType].Assembly.GetCustomAttributes() |
    ? { $_ -is [System.Management.Automation.DynamicClassImplementationAssemblyAttribute] } |
    % { $_.ScriptFile }

### CODE REVIEW

File: src/System.Management.Automation/engine/parser/PSType.cs [hunk 1]
```diff
@@ -1100,30 +1102,42 @@ internal void DefineEnum()
...
+            if (string.IsNullOrEmpty(scriptFile)) {
+                yield return new CustomAttributeBuilder(ctor, emptyArgs);
+                yield break;
+            }
+
+            var propertyInfo = new PropertyInfo[] {
+                typeof(DynamicClassImplementationAssemblyAttribute).GetProperty(nameof(DynamicClassImplementationAssemblyAttribute.ScriptFile)) };
+            var propertyArgs = new object[] { scriptFile };
+            var fieldInfoList = Utils.EmptyArray<FieldInfo>();
+            var fieldArgs = Utils.EmptyArray<object>();
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/parser/PSType.cs]: `fieldInfoList` and `fieldArgs` are not used anywhere.

File: src/System.Management.Automation/engine/parser/PSType.cs [hunk 2]
```diff
@@ -1100,30 +1102,42 @@ internal void DefineEnum()
             }
         }
 
-        private static IEnumerable<CustomAttributeBuilder> GetAssemblyAttributeBuilders()
+        private static IEnumerable<CustomAttributeBuilder> GetAssemblyAttributeBuilders(string scriptFile)
         {
-            yield return new CustomAttributeBuilder(typeof(DynamicClassImplementationAssemblyAttribute).GetConstructor(Type.EmptyTypes), s_emptyArgArray);
+            var ctor = typeof(DynamicClassImplementationAssemblyAttribute).GetConstructor(Type.EmptyTypes);
+            var emptyArgs = Utils.EmptyArray<object>();
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/parser/PSType.cs]: You call `Utils.EmptyArray<object>()` multiple times, why not just use `s_emptyArgArray`?

File: test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1 [hunk 3]
```diff
@@ -688,6 +688,14 @@ Describe 'Type building' -Tags "CI" {
             ++$a::a | Should Be 2
         }
     }
+
+    It 'should get the script from a class type' {
+        class C {}
+
+        $a = [C].Assembly.GetCustomAttributes($false).Where{
+            $_ -is [System.Management.Automation.DynamicClassImplementationAssemblyAttribute]}
+        $a.ScriptFile | Should BeExactly (& { $MyInvocation.ScriptName })
```
Comment (daxian-dbw) [inline test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1]: Maybe replace `(& { $MyInvocation.ScriptName })` with `$PSCommandPath`?

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/parser/PSType.cs (hunk 2 above)
Comment (lzybkr) [reply src/System.Management.Automation/engine/parser/PSType.cs]: I'll remove s_emptyArgArray - it's probably not better to store 2 references to the same array, `Utils.EmptyArray` already has a static reference.

File: src/System.Management.Automation/engine/parser/PSType.cs (hunk 1 above)
Comment (lzybkr) [reply src/System.Management.Automation/engine/parser/PSType.cs]: Thanks, cut and paste I guess.

File: test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1 (hunk 3 above)
Comment (lzybkr) [reply test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1]: Uh, sure. My excuse - I've never used it before and was too lazy to look for it. :)

**Comment by SteveL-MSFT** [comment]
This is great, but the usage seems too complex to remember.  Can we use ETS to just add a ScriptFile property making it more discoverable?

**Comment by lzybkr** [comment]
@SteveL-MSFT - I've only seen 1 request for this ability - from the author of ISESteroids - so ease of use doesn't seem too critical. I suppose we could use a `NoteProperty` on the assembly instance, but that is harder to access from C#.

**Review by daxian-dbw — APPROVED** [review]

**Comment by SteveL-MSFT** [comment]
@lzybkr we can wait for customer request before adding more.  Thanks

**Review by iSazonov — APPROVED** [review]
Leave a comment

**Pull request merged**